src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,130 @@
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
-
+        T[] temporaryArray = (T[]) new Object[size];
+        System.arraycopy(elementData, 0, temporaryArray, 0, size);
+        if (size == currentCapacity) {
+            resize();
+        }
+        isIndexValid(index, "ADD");
+        elementData[index] = value;
+        System.arraycopy(temporaryArray,
+                index,
+                elementData,
+                index + 1,
+                size - index);
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        T[] temporaryArray = (T[]) new Object[list.size()];
+        for (int i = 0; i < temporaryArray.length; i++) {
+            temporaryArray[i] = list.get(i);
+        }
+        System.arraycopy(temporaryArray,
+                0,
+                elementData,
+                size,
+                list.size());
+        size += list.size();
     }
 
     @Override
     public T get(int index) {
-        return null;
+        isIndexValid(index, "noADD");
+        return elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        isIndexValid(index, "noADD");
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        isIndexValid(index, "noADD");
+        T oldRecord = elementData[index];
+        System.arraycopy(elementData,
+                index + 1,
+                elementData,
+                index,
+                size - index - 1);
+        size--;
+        return oldRecord;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        int temporaryIndex = -1;
+        T oldRecord = (T) new Object();
+        for (int i = 0; i < size; i++) {
+            if (elementData[i] == element
+                    || (elementData[i] != null
+                    && elementData[i].equals(element))) {
+                temporaryIndex = i;
+                oldRecord = elementData[i];
+                break;
+            }
+        }
+        if (temporaryIndex != -1) {
+            System.arraycopy(elementData,
+                    temporaryIndex + 1,
+                    elementData,
+                    temporaryIndex,
+                    size - temporaryIndex - 1);
+            size--;
+            return oldRecord;
+        }
+        throw new NoSuchElementException("Element not find");
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void resize() {
+        currentCapacity *= GROW_COEFFICIENT;
+        T[] temporaryArray = (T[]) new Object[size];
+        System.arraycopy(elementData, 0, temporaryArray, 0, size);
+        elementData = (T[]) new Object[currentCapacity];
+        System.arraycopy(temporaryArray, 0, elementData, 0, size);
```

Comment:
we can make it without `temporaryArray` and with only one `System.arraycopy` method call